libs/json/types.go

```diff
@@ -0,0 +1,108 @@
+package json
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"sync"
+)
+
+var (
+	// typeRegistry contains globally registered types for JSON encoding/decoding.
+	typeRegistry = newTypes()
+)
+
+// RegisterType registers a type for Amino-compatible interface encoding in the global type
+// registry. These types will be encoded with a type wrapper `{"type":"<type>","value":<value>}`
+// regardless of which interface they are wrapped in (if any). If the type is a pointer, it will
+// still be valid both for value and pointer types, but decoding into an interface will generate
+// the a value or pointer based on the registered type.
+//
+// Should only be called in init() functions, as it panics on error.
+func RegisterType(_type interface{}, name string) {
```

Comment:
On first sight I was wondering if it is bad to "lose" the connection between the registered interfaces and concrete types but the interface doesn't really matter and the connection will still remain visible in the code via the cdc.RegisterXXX methods in the code.

In rust, these registered interfaces + implementations best are represented via a an enum (enum equiv. to interface and enum variants equiv. to concrete impls). 
e.g. https://github.com/informalsystems/tendermint-rs/blob/cf0763e8a5518e35f9155aeee87ce9093fb833e1/tendermint/src/rpc/event_listener.rs#L112-L134

Reply:
For JSON we never really use the interface for anything, and I'd ideally like to strip this down to the absolute minimum. I think there's only a handful of registered types which will remain (mainly the `Evidence` interface and encryption keys), hopefully we can throw this entire package out when we clean up the RPC layer.